### PR TITLE
Biometrics for unlocking from visual lock

### LINKED ISSUE #35: Add support for iOS biometric APIs to unlock app data
When accessing Lockbox, I want to be able to use Touch or Face ID instead of my password for security and convenience.

## Visual Design
**InVision**
Touch ID: https://mozilla.invisionapp.com/share/XCEV238ZD#/288717368_Welcome_-_Returning_User_-_TouchID
Touch ID w/ prompt: https://mozilla.invisionapp.com/share/XCEV238ZD#/267560531_Welcome_-_Returning_User_-_Touch_ID
Face ID: https://mozilla.invisionapp.com/share/XCEV238ZD#/267560535_Welcome_-_Returning_User
Face ID w/ prompt: https://mozilla.invisionapp.com/share/XCEV238ZD#/288719841_Welcome_-_Returning_User_-_Face_ID

**Zeplin**
Touch ID: https://zpl.io/V0ZW5xO
Touch ID w/ prompt: https://zpl.io/2Ewwx5M
Face ID: https://zpl.io/aRNL9Ln
Face ID w/ prompt: https://zpl.io/b6leoDd

tasks: 
- [x] launch biometric auth from `WelcomeView`
- [x] successful auth dispatches unlock action

Depends on #171

### PR DESCRIPTION
**Pull request by sashei**
resolves #35

### CODE REVIEW

**Comment by devinreams** [comment]
Quick notes here:

1. BuddyBuild shows failing ❌ but shouldn't be. @sashei is on it...
2. Latest on `master` has resulted in merge conflicts ⚠️ 
3. @jhugman tagged for review and is on it this week before end-of-sprint. 👟 

---

@sashei I tested locally with an iPhone 7 and Touch ID worked great! ✅

I also tried an iPhone X and Face ID prompted me for a passcode instead of a face. Is that expected? ⚠️ 


screen shot 2018-04-17 at 12 37 05 pm

File: lockbox-ios/Presenter/WelcomePresenter.swift [hunk 1]
```diff
@@ -59,10 +73,16 @@ class WelcomePresenter {
                     .disposed(by: self.disposeBag)
 
             Observable.combineLatest(lockedObservable, biometricsObservable)
-                    .map { LockedEnabled(appLocked: $0.0, biometricsEnabled: $0.1) }
+                    .map {
+                        LockedEnabled(appLocked: $0.0, biometricsEnabled: $0.1)
```
Comment (devinreams) [inline lockbox-ios/Presenter/WelcomePresenter.swift]: @sashei I think I was able to bypass the "Sign Out" (from Settings) lock:

1. open Settings > Sign Out
2. see "Welcome" screen showing locked, as expected ✅ 
3. Close and re-open app
4. See "Welcome" screen showing locked, as expected ✅ 
5. Force close / kill app and re-open
6. See the item list view, not prompted by "Welcome" to unlock ⚠️

Comment (sashei) [reply lockbox-ios/Presenter/WelcomePresenter.swift]: oooo good catch Devin, I think this is actually in the `RootPresenter`.... I will address.

**Comment by sashei** [comment]
The issue Devin is having with `FaceID` is because we need to request permission before accessing `FaceID` on the device (but not `TouchID`!). We can provide a description of what we will be using `FaceID` for to be shown in the permissions prompt -- @changecourse @sandysage @nickbrandt do we have a preference on what that description should be?

File: lockbox-ios/Presenter/WelcomePresenter.swift [hunk 2]
```diff
@@ -32,23 +33,36 @@ class WelcomePresenter {
...
+        self.settingActionHandler = settingActionHandler
+        self.userInfoStore = userInfoStore
         self.userDefaults = userDefaults
+        self.biometryManager = biometryManager
     }
 
     func onViewReady() {
-        let biometricButtonText = LAContext.usesFaceId ? Constant.string.signInFaceID : Constant.string.signInTouchID
-        let biometricImageName = LAContext.usesFaceId ? "face" : "fingerprint"
+        let biometricsAvailable = self.biometryManager.usesFaceID || self.biometryManager.usesTouchID
```
Comment (jhugman) [inline lockbox-ios/Presenter/WelcomePresenter.swift]: Move this to the `biometryManager`.

`lazy var usesBiometrics: Bool = usesFaceID ||  usesTouchID`

File: lockbox-ios/Presenter/WelcomePresenter.swift [hunk 3]
```diff
@@ -59,10 +73,16 @@ class WelcomePresenter {
...
+                        LockedEnabled(appLocked: $0.0, biometricsEnabled: $0.1)
+                    }
                     .distinctUntilChanged()
                     .map { latest -> Bool in
-                        !latest.appLocked ? true : !latest.biometricsEnabled
+                        if !biometricsAvailable {
+                            return true
+                        }
+
+                        return !latest.appLocked ? true : !latest.biometricsEnabled
```
Comment (jhugman) [inline lockbox-ios/Presenter/WelcomePresenter.swift]: This logic looks… convoluted, or unreadable. 

I _think_ this is equivalent:

```swift
return !biometricsAvailable || !latest.appLocked || !latest.biometricsEnabled
```

**Comment by sashei** [comment]
Ok, I got BuddyBuild running with a set of specs pended... I would like to figure out how to fix these specs, but filed a followup issue in the interest of keeping coverage up / testing important presentation code: #280

**Pull request merged**